pubsub/psclient/client.go

```diff
@@ -1,97 +1,347 @@
+		}
+
+		resp, err := c.receiveMsg()
+		if err != nil {
+			if pstypes.IsIOTimeoutErr(err) {
+				continue
+			}
+			log.Errorf("ReceiveMsg failed: %v", err)
+			return
+		}
+
+		msg, err := DecodeMsg(resp)
+		if err != nil {
+			log.Errorf("Failed to decode message: %v", err)
+			continue
+		}
+
+		switch m := msg.(type) {
+		case *pstypes.ResponseMessage:
+			log.Debugf("Response to %s request ID=%d received. Success = %v. Data: %v",
+				m.RequestEventId, m.RequestId, m.Success, m.Data)
+			respChan := c.responseChan(m.RequestId)
+			if respChan == nil {
+				log.Errorf("receiver failed to find request ID %d", m.RequestId)
+				continue
+			}
+			go func() {
+				respChan <- m // unbuffered
+				close(respChan)
+			}()
+			continue
+		case string:
+			// generic "message"
+			log.Debugf("Message (%s): %s", resp.EventId, m)
+		case int:
+			// e.g. "ping"
+			log.Debugf("Message (%s): %d", resp.EventId, m)
+		case *exptypes.WebsocketBlock:
+			log.Debugf("Message (%s): WebsocketBlock(hash=%s)", resp.EventId, m.Block.Hash)
+		case *exptypes.MempoolShort:
+			t := time.Unix(m.Time, 0)
+			log.Debugf("Message (%s): MempoolShort(numTx=%d, time=%v)",
+				resp.EventId, m.NumAll, t)
+		case *pstypes.TxList:
+			log.Debugf("Message (%s): TxList(len=%d)", resp.EventId, len(*m))
+		case *pstypes.AddressMessage:
+			log.Debugf("Message (%s): AddressMessage(address=%s, txHash=%s)",
+				resp.EventId, m.Address, m.TxHash)
+		default:
+			log.Debugf("Message of type %v unhandled.", resp.EventId)
+			continue
+		}
+
+		c.recvMsgChan <- &ClientMessage{
+			EventId: resp.EventId,
+			Message: msg,
+		}
+	}
+}
+
+func (c *Client) send(msg []byte) error {
+	c.sendMtx.Lock()
+	defer c.sendMtx.Unlock()
+	_ = c.SetWriteDeadline(time.Now().Add(c.writeTimeout))
+	_, err := c.Write(msg)
+	return err
+}
+
+func (c *Client) newRequestID() int64 {
+	c.reqMtx.Lock()
+	reqID := c.nextRequestID
+	c.nextRequestID++
+	c.reqMtx.Unlock()
+	return reqID
+}
+
+func (c *Client) responseChan(reqID int64) chan *pstypes.ResponseMessage {
+	c.reqMtx.Lock()
+	defer c.reqMtx.Unlock()
+	return c.requests[reqID]
+}
+
+func (c *Client) newResponseChan() (chan *pstypes.ResponseMessage, int64) {
+	c.reqMtx.Lock()
+	reqID := c.nextRequestID
+	c.nextRequestID++
+	respChan := make(chan *pstypes.ResponseMessage)
+	c.requests[reqID] = respChan
+	c.reqMtx.Unlock()
+	return respChan, reqID
+}
+
+func (c *Client) deleteRequestID(reqID int64) {
+	c.reqMtx.Lock()
+	delete(c.requests, reqID)
+	c.reqMtx.Unlock()
+}
+
 // Subscribe sends a subscribe type WebSocketMessage for the given event name
 // after validating it. The response is returned.
-func (c *Client) Subscribe(event string) (*pstypes.WebSocketMessage, error) {
+func (c *Client) Subscribe(event string) (*pstypes.ResponseMessage, error) {
 	// Validate the event type.
 	_, _, ok := pstypes.ValidateSubscription(event)
 	if !ok {
 		return nil, fmt.Errorf("invalid subscription %s", event)
 	}
 
+	respChan, reqID := c.newResponseChan()
```

Comment:
The channel on which the response should be received, and a new unique request ID for generating the request message.